ec-gpu/src/lib.rs

```diff
@@ -1,22 +1,30 @@
-/// Describes how to generate the elliptic curve operations for
-/// - `Scalar`
-/// - `Fp`
-/// - `Fp2`
-/// - `G1`
-/// - `G2`
-pub trait GpuEngine {
-    type Scalar: GpuField;
-    type Fp: GpuField;
+/// The name that is used in the GPU source code to identify the item that is used.
+pub trait GpuName {
+    /// The name of the item.
+    ///
+    /// This name is used in the source code that is generated for the GPU. It should be globally
+    /// unique so that you don't collide with other libraries. It must *not* contain spaces. For
+    /// fields it is best to use a combination of the curve you are using and the field name. For
+    /// example `bls12_381_fp`, `bls12_381_scalar`, or `pallas_fp`.
+    fn name() -> String;
```

Comment:
This global uniqueness requirement is going to be very hard to enforce; in fact, this documentation recommends a strategy for naming fields that is guaranteed to cause collisions across different implementations of the same field.

However, we know that within a particular binary application, each type will have a unique [`TypeId`](https://doc.rust-lang.org/std/any/struct.TypeId.html). Would it be possible to generate the GPU source code from something based on that identifier, as derived from `Self`? The `TypeId` is allowed to change between compilation runs (see [this issue](https://github.com/rust-lang/rust/issues/61553) for some context), so it wouldn't work if the expectation is that the generated GPU source code can be compiled once and reused indefinitely, but if the source is being generated and compiled live (and thus is already subject to the `rustc -Cmetadata` flag) then this should both work, and mean downstream crates don't need to do anything (as the changes to `GpuField` could then be entirely non-breaking, either with this being a default method on `GpuField` or by inverting the trait bounds so `GpuName: GpuField` and providing a blanket implementation of `GpuName`).

Reply:
I thought about using `TypeId`, but then thought it might be too brittle. Practically I wouldn't expect that there are bazillions of implementations. I even thought one might perhaps even use a non-unique name intentionally. E.g. you might even want to create the same source for differently implemented fields (thought I'm not really sure if that's a use case). I liked the idea of keeping things pretty straight forward and easy to debug. Hence I think explicit names are better.

Input from other folks would be appreciated.

Even if we would use `TypeId` it would still be a breaking change for downstream crates, as there are cases where `GpuName` is implemented for things that aren't fields. So the trait bound cannot be flipped.

Reply:
> Practically I wouldn't expect that there are bazillions of implementations.

That may be true, but it only takes two to collide in order for the documented requirement to fail. And there are no clearly documented consequences for a collision.

> I even thought one might perhaps even use a non-unique name intentionally. E.g. you might even want to create the same source for differently implemented fields (thought I'm not really sure if that's a use case).

This is where it would be _very_ useful to have more documentation about how this function will (or should) be used. If the contents of this method are not being tightly constrained, then it falls to the consumer of the method to have a well-defined usage contract, so that people implementing this trait know what they are implementing it against.

> I liked the idea of keeping things pretty straight forward and easy to debug. Hence I think explicit names are better.

Definitely reasonable, and also why I'd prefer human-meaningful names over compiler-generated identifiers. If debuggability were the only concern, then [`std::any::type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html) would be perfect for this; sadly it doesn't guarantee to include the module path, so it doesn't claim uniqueness.

Maybe the recommended implementation for this method should be:
```rust
    fn name() -> String {
        format!("{}::{}", module_path!(), std::any::type_name::<Self>())
    }
```
(or something more involved that strips the module path if present from the output of `type_name`). This could even be provided as a macro (used like e.g. `ec_gpu::name!(MyType);`) so that most people use the non-colliding default (while people with more convoluted use cases can implement the trait directly).

> Even if we would use `TypeId` it would still be a breaking change for downstream crates, as there are cases where `GpuName` is implemented for things that aren't fields. So the trait bound cannot be flipped.

Good to know.

Reply:
Thanks @str4d for all the input. I played a bit around with those ideas and pushed a new commit. Now the name is auto-generated and it actually isn't too bad. In case of BLS381-12, it's names like `ec_gpu_blstrs__fp__Fp`, `ec_gpu_blstrs__scalar__Scalar` or `ec_gpu_blstrs__g1__G1Affine`, which are still reasonable. I've also updated the documentation that you shouldn't implement the name yourself.

I thought it would be a good idea to auto-implement it for the `GpuField`. I don't know if it's still a breaking change (I guess it still is), but at least it will need less code changes. I've also updated the `blstrs` PR so that folks can see what changes would be needed https://github.com/filecoin-project/blstrs/pull/43. I think the `pasta_curves` trait wouldn't need any code changes.

Reply:
I forgot to add that `G1` and `G2` (https://github.com/filecoin-project/blstrs/pull/43/files#diff-a81fd3c080fd012c0c9155e68a28c5ef92bafcd33eced7847bbb40c5a7dcc116R841) are an example where I need a `GpuName` which aren't fields. I need it in order to name the Multiexp/MSM kernel correctly.